### DISCUSSION TITLE
`Promise.all()` within interactive transactions

### DISCUSSION THREAD

**Discussion by michaelhays** [prisma/prisma#21550] [Q&A] [answered]
### Question

Hello, I'm using [Prisma interactive transactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions) with my MySQL database, and saw some surprising results with the usage of `Promise.all()`. It looks like requests made within `Promise.all()` are executed serially if wrapped in an interactive transaction.

Is this expected behavior?

Here are some benchmarks for a simple schema:

```ts
await Promise.all([
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
  prisma.user.findMany(),
])
```

```sql
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 520ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 524ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 527ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 533ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 534ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 536ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 575ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 578ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 579ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 625ms
```

^ **Total duration: 639.29ms**

And when wrapped in a transaction:

```ts
await prisma.$transaction(async (prisma) => {
  await Promise.all([
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
    prisma.user.findMany(),
  ])
})
```

```sql
prisma:query -- Implicit "BEGIN" query via underlying driver
Duration: 0ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 206ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 147ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 127ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 191ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 118ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 225ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 142ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 271ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 326ms
prisma:query SELECT `db`.`User`.`id` FROM `db`.`User` WHERE 1=1
Duration: 152ms
prisma:query -- Implicit "COMMIT" query via underlying driver
Duration: 0ms
```

^ **Total duration: 2391.83ms**

I tested with the different [isolation levels](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#supported-isolation-levels) as well, and got the same results.

**Comment by janpio** [comment, marked as answer]
Yes, I would say this is expected.

One transaction means that all queries inside it have to be run on the same connection. A database connection can only ever execute 1 query at a time. As a query blocks the connection while it is doing its work, putting a transaction into `Promise.all` effectively means asking Prisma to run one query after another.

Do you agree?

**Reply by michaelhays** [reply]
Thanks for the response! Yes, that seems right, I thought it might have been something like this.

It may be worth mentioning this in the [interactive transactions docs](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions), since it may come as unexpected for people less familiar with the nuances of database queries/transactions, and can have pretty significant performance implications.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
